core/src/main/scala/chisel3/experimental/Const.scala

```diff
@@ -0,0 +1,15 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.experimental
+
+import chisel3._
+
+/** Create a constant type in FIRRTL, which is guaranteed to take a single
+  * constant value.
+  */
+object Const {
+  def apply[A <: Data](a: A): A = {
+    a.isConst = true
+    a
+  }
+}
```

Comment:
Chisel APIs need to give the illusion of immutability even if they actually use mutation under the hood. We typically do this by cloning the argument, mutating the value on the clone, and then returning the clone although we have optimized this a bit to only clone when we need to. Take a look at how `Input`, `Output`, and `Flipped` work (which are setting the direction while pretending to not mutate anything).

Reply:
Are you referring to the code here? https://github.com/chipsalliance/chisel3/blob/4194e10697ce501a79da2ddad323f2e3dced830a/core/src/main/scala/chisel3/Data.scala#L59

Is it ok to use `cloneType` for this? I am getting the following when using cloneType in the `Bundle` case:
```
chisel3.package$ExpectedHardwareException: 'FooBundle' must be hardware, not a bare Chisel type. Perhaps you forgot to wrap it in Wire(_) or IO(_)?
```

Reply:
Yes something like that. Yes you in fact _need_ to clone the thing at times, that's the point. Instead of `cloneType` though you should use `cloneTypeFull` which maintains set specified direction. I don't know why you're hitting that exception, but that must be a bug in your code, the point stands that `Const` must clone things (when it needs to as defined by `mustClone` as used in the logic you linked).

Reply:
Ah yes, I see my bug, thanks Jack!